### PR TITLE
Complete Doctest Coverage for GluCat 0.98a1

### LINKED ISSUE #74: Complete C++ doctest coverage
Add more C++ doctests including mocks so that test coverage is as close to 100% as possible.

### PR DESCRIPTION
**Pull request by penguian**
Closes #74 

**Description**
This PR finalizes the migration to the decentralized C++ `doctest` framework and successfully achieves the milestone of **100% function coverage** and **>90% line coverage** for the core GluCat library.

**Key Changes & Achievements:**
* **Doctest Migration Complete:** Fully migrated and translated the legacy PyClical Python doctest suite into native C++ `doctest` `SUBCASE` blocks, deeply integrating them directly into the core library implementation headers (`*_imp.h`).
* **Hardened Edge Cases:** Added comprehensive unit tests targeting complex branching paths across all matrix backends, including Eigen (`matrix_eigen_imp.h`), Armadillo (`matrix_arma_imp.h`), and multi-framed representations (`framed_multi_imp.h`, `matrix_multi_imp.h`).
* **Template Disambiguation:** Fixed various template parameter ambiguities and constructor overload compilation issues uncovered during the coverage hardening process for multiple precision arithmetic types (e.g., `dd_real`, `qd_real`).
* **Documentation Updates:** Bumped the version to `0.98a1` project-wide and updated `NEWS`, `README.md`, and `INSTALL.md` to reflect the new 100% coverage milestone and the completion of the `doctest` transition.

**Changed Components:**
* `glucat/*_imp.h`: Major additions of decentralized test cases.
* `test_doctest/*`: Refactoring and consolidation of test headers.
* `NEWS` / `README.md` / `INSTALL.md`: Release notes and metrics.

### CODE REVIEW

File: glucat/matrix_multi_imp.h [hunk 1]
```diff
@@ -3240,15 +3243,144 @@ TEST_CASE("matrix_multi<Scalar_T, LO, HI, Tune_P>") {
...
+    mm_t f_inf(std::numeric_limits<T>::infinity());
+    
+    CHECK(f_nan.isnan());
+    CHECK(f_inf.isinf());
+  }
+
+  SUBCASE("File I/O") {
+    mm_t f("1+{1}");
+    f.write("Test prefix"); 
+
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_multi_imp.h]: This behavior is intentional. `doctest` captures standard output during successful test execution and only displays it if the test fails. The "Test prefix" provides valuable context for debugging during failure analysis without cluttering CI logs during normal runs.

File: glucat/framed_multi_imp.h [hunk 2]
```diff
@@ -2921,7 +3007,124 @@ TEST_CASE("framed_multi<Scalar_T, LO, HI, Tune_P>") {
...
+  SUBCASE("Extended I/O") {
+    fm_t f("1+{1}");
+    f.write("Test prefix"); // Mostly for coverage
+    
+    std::ofstream ofs("test_io.txt");
+    f.write(ofs, "File prefix");
+    ofs.close();
+    CHECK(std::filesystem::exists("test_io.txt"));
+    std::filesystem::remove("test_io.txt");
+  }
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/framed_multi_imp.h]: Agreed. I have refactored the File I/O tests to use `std::filesystem::temp_directory_path()` to generate unique paths and implemented RAII-based cleanup to ensure test artifacts are removed automatically, even if the test fails or aborts.

File: glucat/matrix_eigen.h [hunk 3]
```diff
@@ -109,7 +109,7 @@ namespace glucat { namespace matrix
 
     // Constructor from eigen_sparse_wrapper
     template< typename Other_Scalar_T >
-    explicit eigen_matrix_wrapper(const eigen_sparse_wrapper<Other_Scalar_T>& other);
+    eigen_matrix_wrapper(const eigen_sparse_wrapper<Other_Scalar_T>& other);
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_eigen.h]: Agreed. After further review, I have restored the `explicit` keyword. The potential for accidental, high-cost memory allocations from implicit sparse-to-dense conversions is a significant safety risk that outweighs the minor ergonomic benefits of an implicit conversion.

File: glucat/clifford_algebra_imp.h [hunk 4]
```diff
@@ -1742,12 +1742,121 @@ TEST_CASE("clifford_algebra_imp (generic templates)") {
...
+    CHECK(approx_equal(a.involute().involute(), a));
+    CHECK(approx_equal(a.reverse().reverse(), a));
+    
+    // inv() for non-scalar
+    mm_t e1(index_set_t(1), 1.0, index_set_t(1), true);
+    mm_t inv_test = mm_t(1.0) + e1 * 0.5;
+    try {
+      mm_t inv_a = inv_test.inv();
+      CHECK(approx_equal(inv_a * inv_test, mm_t(1.0)));
+    } catch (...) {}
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/clifford_algebra_imp.h]: Agreed. I have replaced the empty `try/catch` block with `CHECK_NOTHROW(inv_test.inv())`. This ensures that any unexpected exception is correctly flagged as a test failure rather than being silently ignored.

[the remark this answers is not in the document]
Comment (penguian) [reply glucat/clifford_algebra_imp.h]: Fixed. The redundant `using index_set_t` declaration has been removed to resolve the compilation error.

File: glucat/matrix_multi_imp.h [hunk 5]
```diff
@@ -3141,7 +3144,7 @@ TEST_CASE("matrix_multi<Scalar_T, LO, HI, Tune_P>") {
 
   SUBCASE("Geometric algebra identities (random)") {
     using index_set_t = mm_t::index_set_t; using index_set_t = mm_t::index_set_t; index_set_t frm = index_set_t();
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_multi_imp.h]: Fixed. The redundant `using index_set_t` declaration has been removed to resolve the compilation error.

File: glucat/matrix_eigen_imp.h [hunk 6]
```diff
@@ -1600,69 +1613,213 @@ namespace glucat { namespace matrix
 #include <doctest.h>
 #include <iostream>
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_eigen_imp.h]: Fixed. Added `#include <sstream>` within the `#ifdef GLUCAT_DOCTEST` block to ensure the test code is self-contained.

File: glucat/matrix_multi_imp.h [hunk 7]
```diff
@@ -3240,15 +3243,144 @@ TEST_CASE("matrix_multi<Scalar_T, LO, HI, Tune_P>") {
...
+  SUBCASE("File I/O") {
+    mm_t f("1+{1}");
+    f.write("Test prefix"); 
+    
+    std::ofstream ofs("test_io_mm.txt");
+    f.write(ofs, "File prefix");
+    ofs.close();
+    CHECK(std::filesystem::exists("test_io_mm.txt"));
+    std::filesystem::remove("test_io_mm.txt");
+  }
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_multi_imp.h]: Agreed. Refactored these tests to use unique temporary paths with RAII cleanup, consistent with the changes made in `framed_multi_imp.h`.

File: glucat/framed_multi_imp.h [hunk 8]
```diff
@@ -2921,7 +3007,124 @@ TEST_CASE("framed_multi<Scalar_T, LO, HI, Tune_P>") {
...
+    
+    CHECK(f_nan.isnan());
+    CHECK(!f_nan.isinf());
+    CHECK(f_inf.isinf());
+    CHECK(!f_inf.isnan());
+  }
+
+  SUBCASE("Extended I/O") {
+    fm_t f("1+{1}");
+    f.write("Test prefix"); // Mostly for coverage
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/framed_multi_imp.h:3101]: Consistent with the response to the first conversation, this is intentional as `doctest` captures output by default. It is preserved as useful diagnostic information that only appears in the event of a test failure.

File: test_doctest/test_doctest.h [hunk 9]
```diff
@@ -0,0 +1,134 @@
+#ifndef _GLUCAT_TEST_DOCTEST_TEST_DOCTEST_H
+#define _GLUCAT_TEST_DOCTEST_TEST_DOCTEST_H
+
+#include <doctest.h>
+#include "glucat/glucat.h"
+#include <iostream>
+#include <fstream>
```
[the remark this answers is not in the document]
Comment (penguian) [reply test_doctest/test_doctest.h]: Fixed. Added `#include <limits>` to the test header to ensure portability and reduce fragility across different compiler versions.

File: glucat/matrix_arma_imp.h [hunk 10]
```diff
@@ -1497,76 +1515,204 @@ namespace glucat { namespace matrix
 #include <doctest.h>
 #include <iostream>
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_arma_imp.h]: Fixed. Added `#include <sstream>` within the `#ifdef GLUCAT_DOCTEST` block to ensure the test code is self-contained.

File: glucat/errors_imp.h [hunk 11]
```diff
@@ -77,17 +77,44 @@ namespace glucat
...
 
-  error<dummy> e2("message2");
-  CHECK(e2.classname() == "dummy");
+  SUBCASE("Output verification (std::cerr redirection)") {
+    error<dummy_class> e("context", "message");
+    
+    // Redirect cerr to a stringstream
+    std::stringstream buffer;
+    std::streambuf* old_cerr = std::cerr.rdbuf(buffer.rdbuf());
+
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/errors_imp.h]: Fixed. Added the missing `#include <sstream>` to ensure the error handling tests are self-contained and independent of transitive includes.

File: glucat/matrix_multi_imp.h [hunk 12]
```diff
@@ -42,6 +42,8 @@
 #include <array>
 #include <iostream>
 #include <fstream>
+#include <filesystem>
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_multi_imp.h]: I moved the `#include <filesystem>` into the guarded test block.

File: glucat/framed_multi_imp.h [hunk 13]
```diff
@@ -42,6 +42,8 @@
 
 #include <sstream>
 #include <fstream>
+#include <filesystem>
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/framed_multi_imp.h]: I moved the `#include <filesystem>` into the guarded test block.

File: glucat/errors_imp.h [hunk 14]
```diff
@@ -77,17 +77,45 @@ namespace glucat
 }
 
 #ifdef GLUCAT_DOCTEST
+#include <sstream>
+struct dummy_class { 
+  static constexpr std::string_view name = "dummy_class";
+  static std::string_view classname() { return name; } 
+};
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/errors_imp.h]: I added `#include <string_view>` to the guarded test bock.

File: glucat/clifford_algebra.h [hunk 15]
```diff
@@ -145,47 +145,47 @@ namespace glucat
...
+    auto outer_pow   (int m) const                     -> multivector_t ;\
+    auto frame       () const                          -> index_set_t   ;\
+    auto grade       () const                          -> index_t             ;\
+    auto operator[]  (const index_set_t ist) const     -> Scalar_T            ;\
+    auto operator()  (index_t grade) const             -> multivector_t ;\
+    auto scalar      () const                          -> Scalar_T            ;\
+    auto pure        () const                          -> multivector_t ;\
+    auto even        () const                          -> multivector_t ;\
+    auto odd         () const                          -> multivector_t ;\
+    auto vector_part () const                          -> vector_t      ;\
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/clifford_algebra.h]: Done.

File: glucat/framed_multi_imp.h [hunk 16]
```diff
@@ -2911,6 +2954,50 @@ TEST_CASE("framed_multi<Scalar_T, LO, HI, Tune_P>") {
...
+    // Trigger exception path for constructor
+    index_set_t invalid_ist(2);
+    CHECK_THROWS_AS(fm_t(invalid_ist, crd, frm, false), glucat_error);
+  }
+
+  SUBCASE("Fast Path (32 terms)") {
+    using index_set_t = fm_t::index_set_t;
+    using scalar_t = fm_t::scalar_t;
+    index_set_t large_frm;
+    for(int i=1; i<=6; ++i) large_frm |= index_set_t(i); // 64 dimensions
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/framed_multi_imp.h]: Done.

File: glucat/matrix_arma_imp.h [hunk 17]
```diff
@@ -1495,78 +1513,238 @@ namespace glucat { namespace matrix
 } }
 #ifdef GLUCAT_DOCTEST
 #include <doctest.h>
+#include <sstream>
 #include <iostream>
+#include <sstream>
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_arma_imp.h:1516]: Done.

File: glucat/matrix_eigen_imp.h [hunk 18]
```diff
@@ -1599,70 +1612,248 @@ namespace glucat { namespace matrix
 #ifdef GLUCAT_DOCTEST
 #include <doctest.h>
 #include <iostream>
+#include <sstream>
+#include <sstream>
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_eigen_imp.h:1616]: Done.

File: glucat/framed_multi_imp.h [hunk 19]
```diff
@@ -2921,7 +3008,161 @@ TEST_CASE("framed_multi<Scalar_T, LO, HI, Tune_P>") {
...
+  SUBCASE("I/O and Formatting") {
+    fm_t f(1.23456789);
+    std::ostringstream oss;
+    oss << std::setprecision(4) << std::fixed << f;
+    CHECK(oss.str().find("1.2346") != std::string::npos);
+    
+    oss.str("");
+    oss << std::scientific << f;
+    CHECK(oss.str().find("1.234") != std::string::npos);
+    CHECK(oss.str().find("e+00") != std::string::npos);
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/framed_multi_imp.h:3024]: Done.

File: test_doctest/test_doctest.h [hunk 20]
```diff
@@ -0,0 +1,135 @@
...
```
[the remark this answers is not in the document]
Comment (penguian) [reply test_doctest/test_doctest.h]: Addressed.

File: glucat/errors_imp.h [hunk 21]
```diff
@@ -77,17 +77,46 @@ namespace glucat
...
+    error<dummy_class> e("context", "message");
+
+    // Redirect cerr to a stringstream
+    std::stringstream buffer;
+    std::streambuf* old_cerr = std::cerr.rdbuf(buffer.rdbuf());
+
+    e.print_error_msg();
+
+    // Restore cerr
+    std::cerr.rdbuf(old_cerr);
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/errors_imp.h]: Addressed.

File: glucat/framed_multi_imp.h [hunk 22]
```diff
@@ -2901,27 +2945,225 @@ TEST_CASE("framed_multi<Scalar_T, LO, HI, Tune_P>") {
...
+    CHECK(!f_inf.isnan());
+  }
+
+  SUBCASE("Extended I/O") {
+    fm_t f("1+{1}");
+    f.write("Test prefix"); // Mostly for coverage
+
+    auto temp_path = std::filesystem::temp_directory_path() / ("test_io_" + std::to_string(std::chrono::system_clock::now().time_since_epoch().count()) + ".txt");
+    struct Cleanup { std::filesystem::path p; ~Cleanup() { if(std::filesystem::exists(p)) std::filesystem::remove(p); } } cleanup{temp_path};
+
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/framed_multi_imp.h:3105]: Addressed.

File: glucat/matrix_multi_imp.h [hunk 23]
```diff
@@ -3240,15 +3244,146 @@ TEST_CASE("matrix_multi<Scalar_T, LO, HI, Tune_P>") {
...
+    CHECK(f_inf.isinf());
+  }
+
+  SUBCASE("File I/O") {
+    mm_t f("1+{1}");
+    f.write("Test prefix");
+
+    auto temp_path = std::filesystem::temp_directory_path() / ("test_io_mm_" + std::to_string(std::chrono::system_clock::now().time_since_epoch().count()) + ".txt");
+    struct Cleanup { std::filesystem::path p; ~Cleanup() { if(std::filesystem::exists(p)) std::filesystem::remove(p); } } cleanup{temp_path};
+
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_multi_imp.h:3337]: Addressed.

File: glucat/matrix_base_imp.h [hunk 24]
```diff
@@ -270,4 +270,69 @@ namespace glucat { namespace matrix
...
+    // - a gap distribution where the wrap-around gap is strictly largest
+    MatrixDbl_T m(5, 5);
+    m.zeros();
+    // Block 1: rot by pi/2 -> eigenvalues {i, -i}
+    m(0,1) = -1.0;
+    m(1,0) = 1.0;
+    // Block 2: rot by 0.6*pi -> eigenvalues {e^i0.6pi, e^-i0.6pi}
+    double theta = 0.6 * 3.14159265358979323846;
+    m(2,2) = std::cos(theta); m(2,3) = -std::sin(theta);
+    m(3,2) = std::sin(theta); m(3,3) = std::cos(theta);
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_base_imp.h:297]: Addressed.

File: glucat/generation_imp.h [hunk 25]
```diff
@@ -293,4 +293,49 @@ namespace glucat { namespace gen
...
+    // Signature (1, 7) -> triggers gen_from_pp4_qm4 (via bott=2 path)
+    wide_index_set_t p1, n7;
+    p1 |= wide_index_set_t(1);
+    for(int i=1; i<=7; ++i) n7 |= wide_index_set_t(-i);
+    WideMultivector_T v_p1n7(p1 | n7, wide_scalar_t(1.0), p1 | n7, true);
+
+    // Signature (10, 0) -> triggers gen_from_pm4_qp4 (via bott=2 path)
+    wide_index_set_t p10;
+    for(int i=1; i<=10; ++i) p10 |= wide_index_set_t(i);
+    WideMultivector_T v_p10(p10, wide_scalar_t(1.0), p10, true);
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/generation_imp.h:336]: Done

File: glucat/matrix_multi.h [hunk 26]
```diff
@@ -159,7 +159,7 @@ namespace glucat
     // Class name used in messages
     static std::string_view classname();
     /// Destructor
-    ~matrix_multi() override = default;
+    ~matrix_multi() = default;
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_multi.h]: Done

File: glucat/framed_multi.h [hunk 27]
```diff
@@ -183,7 +186,7 @@ namespace glucat
     // Class name used in messages
     static auto classname() -> std::string_view;
     /// Destructor
-    ~framed_multi() override = default;
+    ~framed_multi() = default;
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/framed_multi.h]: Done

File: glucat/index_set.h [hunk 28]
```diff
@@ -104,6 +104,24 @@ namespace glucat
...
+    }
+
+    /// Create index set from a range of indices at compile-time
+    template<const index_t BEGIN, const index_t END>
+    static auto from_range() -> index_set_t
+    {
+      static_assert(BEGIN <= END, "index_set::from_range: invalid range [BEGIN, END]");
+      static_assert(BEGIN >= LO && END <= HI, "index_set::from_range: range out of bounds");
+      return index_set_t(index_pair_t(BEGIN, END), true);
+    }
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/index_set.h:123]: Done

File: glucat/matrix_eigen_imp.h [hunk 29]
```diff
@@ -1211,9 +1211,22 @@ namespace glucat { namespace matrix
       return true;
     if (m_outer >= static_cast<matrix_index_t>(mp_mat->outerSize()))
       return false;
-    return m_inner != other.m_inner;
+    return m_inner.index() != other.m_inner.index();
   }
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_eigen_imp.h:1217]: Done

File: glucat/matrix_multi_imp.h [hunk 30]
```diff
@@ -3240,15 +3244,146 @@ TEST_CASE("matrix_multi<Scalar_T, LO, HI, Tune_P>") {
...
+    mm_t m2(is_t("{1}"), T(1.0));
+    mm_t sum = m1 + m2; // Triggers reframe
+    CHECK(sum.frame() == is_t("{1}"));
+    CHECK(sum.scalar() == doctest::Approx(T(1.0)));
+    CHECK(sum[is_t("{1}")] == doctest::Approx(T(1.0)));
+  }
+
+  SUBCASE("Advanced Matrix Algorithms") {
+    mm_t m1(T(1.0), mm_t::index_set_t());
+    mm_t m2 = m1.outer_pow(2);
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_multi_imp.h:3308]: Done

File: glucat/matrix_multi_imp.h [hunk 31]
```diff
@@ -3240,15 +3243,147 @@ TEST_CASE("matrix_multi<Scalar_T, LO, HI, Tune_P>") {
...
   SUBCASE("Exceptions") {
     mm_t m1(1.0);
     // Negative exponent in outer_pow
     CHECK_THROWS(m1.outer_pow(-1));
 
     // Construction with value outside of frame
-    // mm_t is <-32, 32>, index_set_t(33) is out of frame
+    // mm_t is <-8, 8>, index_set_t(9) is out of frame
+    using is_t = mm_t::index_set_t;
+    CHECK_THROWS(mm_t(is_t(9), 1.0, is_t(), false));
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/matrix_multi_imp.h]: Done

File: glucat/framed_multi_imp.h [hunk 32]
```diff
@@ -2901,27 +2945,251 @@ TEST_CASE("framed_multi<Scalar_T, LO, HI, Tune_P>") {
...
     CHECK_THROWS(fm_t("{invalid}"));
-    
+
     // Negative exponent
     CHECK_THROWS(f1.outer_pow(-1));
 
     // Construction with value outside of frame
     using is_t = fm_t::index_set_t;
-    CHECK_THROWS(fm_t(is_t(33), 1.0, is_t(), false));
+    CHECK_THROWS(fm_t(is_t(9), T(1.0), is_t(), false));
```
[the remark this answers is not in the document]
Comment (penguian) [reply glucat/framed_multi_imp.h]: Done

File: NEWS [hunk 33]
```diff
@@ -10,8 +20,6 @@ o Optimized sparse matrix unit() and trace() operations to utilize
   backend-specific diagonal access, avoiding slow manual loops.
 o Integrated a new Doctest-based unit testing suite in test_doctest with 
   comprehensive coverage for all matrix backends.
-o Hardened the test coverage infrastructure by implementing a binary backup 
-  strategy and isolating profiling flags from library detection probes.
 o Restored and verified high-fidelity coverage visibility for QD (quadruple 
   precision) and Armadillo code paths.
```
[the remark this answers is not in the document]
Comment (penguian) [reply NEWS:24]: Addressed

File: README.md [hunk 34]
```diff
@@ -417,8 +419,11 @@ parameters.
...
+* **Template Disambiguation**: Resolved template parameter ambiguities and constructor overload issues in `framed_multi_imp.h` and `matrix_multi_imp.h` for multiple scalar types.
+
 As of GluCat 0.98a0, the following changes have been made:
-* **Coverage Infrastructure Hardening**: Resolved a critical bug where legacy test binaries were deleted before analysis. Implemented a binary backup strategy and isolated profiling flags from library detection probes, restoring full coverage visibility for QD and Armadillo.
 * **Non-Complex Scalar_T Assertion**: Enforced non-complex `Scalar_T` requirements using `static_assert` in `clifford_algebra.h` and `scalar.h` to ensure template robustness and library-wide compatibility.
 * **Sparse Diagonal Optimization**: Optimization of sparse matrix `unit()` and `trace()` operations in Eigen and Armadillo backends to avoid slow manual diagonal write loops.
 * **Return Type Modernization**: Refactored the public API across all core classes and matrix wrappers to use standard return types (e.g., `T func()`) instead of trailing return types (`auto func() -> T`).
```
[the remark this answers is not in the document]
Comment (penguian) [reply README.md:430]: Addressed

**Comment by penguian** [comment]
Merge then tag `master` as `release_0_98_a1`.
